store/postgres/migrations/2024-01-05-170000_ds_corruption_fix_up/up.sql

```diff
@@ -0,0 +1,96 @@
+/*
+
+
+This migration exists to fix up DB inconsistencies caused by a bug that was already fixed in this PR:
+https://github.com/graphprotocol/graph-node/pull/5083
+
+What it does is:
+1. select latest_ethereum_block_number from subgraphs.subgraph_deployment where deployment = '$Qm..';
+2. With that value, check if there are any data sources with a higher block:
+   select count(*) from sgd*.data_sources$ where lower(block_range) > $latest_block;
+3. If there are, then we need to intervene in that subgraph by removing all such data sources:
+   delete from sgd*.data_sources$ where lower(block_range) > $latest_block;
+4. It also unclamps any data sources that need unclamping (even though I don't think we ever clamp them).
+*/
+
+CREATE TEMPORARY TABLE temp_sgd_last_block (
+    deployment_schema text,
+    last_block_from_registry numeric,
+    is_ok boolean,
+    is_upper_limit_too_high boolean
+);
+
+-- collect the latest block number from SG deployment details
+INSERT INTO temp_sgd_last_block (deployment_schema, last_block_from_registry)
+SELECT d.name, latest_ethereum_block_number
+  FROM subgraphs.subgraph_deployment AS sd
+  JOIN deployment_schemas AS d ON sd.deployment = d.subgraph
+ WHERE d.name IN (SELECT relnamespace::regnamespace::name
+                    FROM pg_class
+                   WHERE relname LIKE 'data_sources$%' AND relkind = 'r'
+                 );
+
+-- check if the block numbers in the tables are OK
+-- if not, it can be
+-- - either the lower bound is higher than the last block
+-- - or the lower is fine, but the higher is past the last block
+DO $do$
+DECLARE
+    s text;
+BEGIN
+    FOR s IN SELECT deployment_schema FROM temp_sgd_last_block
+    LOOP
+        EXECUTE format($$
+        UPDATE temp_sgd_last_block
+           SET is_ok = NOT EXISTS (SELECT 1
+                                     FROM %1$I."data_sources$"
+                                    WHERE lower(block_range) > last_block_from_registry
+                                    LIMIT 1)
+                       AND NOT EXISTS (SELECT 1
+                                         FROM %1$I."data_sources$"
+                                        WHERE upper(block_range) > last_block_from_registry 
+                                          AND lower(block_range) <= last_block_from_registry
+                                        LIMIT 1),
```

Comment:
Don't these two conditions just boil down to `where upper(block_range) > last_block_from_registry`? In the first branch if `lower(block_range) > last_block_from_registry`, then the same condition holds with `upper`. Basically, we see if there's a data source that's visible beyond `last_block_from_registry`

Reply:
They do, I just left it like that for expliciteness' sake (to match the explanation above the statement)